tests/test_models.cc

```diff
@@ -10,247 +10,140 @@
- * have been distorted by non-constant noise (heteroscedastic), we then perform
- * cross-validated evaluation of a GaussianProcess which takes that noise into
- * account, and one which is agnostic of the added noise and assert that taking
- * noise into account improves the model.
- */
-TEST(test_models, test_with_target_distribution) {
-  auto dataset = make_heteroscedastic_toy_linear_data();
-
-  auto folds = leave_one_out(dataset);
-  auto model = MakeGaussianProcess().create();
-  EvaluationMetric<Eigen::VectorXd> rmse =
-      evaluation_metrics::root_mean_square_error;
-  auto scores = cross_validated_scores(rmse, folds, model.get());
-  RegressionDataset<double> dataset_without_variance(dataset.features,
-                                                     dataset.targets.mean);
-  auto folds_without_variance = leave_one_out(dataset_without_variance);
+TYPED_TEST(RegressionModelTester, performs_reasonably_on_linear_data) {
+  auto dataset = this->test_case.get_dataset();
+  auto model = this->test_case.get_model();
 
-  auto scores_without_variance =
-      cross_validated_scores(rmse, folds_without_variance, model.get());
+  const auto fit_model = model.get_fit_model(dataset.features, dataset.targets);
+  const auto pred = fit_model.get_prediction(dataset.features);
+  const auto pred_mean = pred.mean();
 
-  EXPECT_LE(scores.mean(), scores_without_variance.mean());
+  double rmse = sqrt((pred_mean - dataset.targets.mean).norm());
+  EXPECT_LE(rmse, 0.5);
 }
 
-TYPED_TEST(RegressionModelTester, cross_validation_variants) {
-  auto dataset = this->creator.get_dataset();
-  auto folds = leave_one_out(dataset);
-  auto model = this->creator.create();
-  EvaluationMetric<Eigen::VectorXd> rmse =
-      evaluation_metrics::root_mean_square_error;
-  auto cv_scores = cross_validated_scores(rmse, folds, model.get());
+TYPED_TEST(RegressionModelTester, test_predict_variants) {
+  auto dataset = this->test_case.get_dataset();
+  auto model = this->test_case.get_model();
 
-  auto loo_indexers = leave_one_out_indexer(dataset);
-  auto loo_predictions =
-      model->template cross_validated_predictions<MarginalDistribution>(
-          dataset, loo_indexers);
+  const auto fit_model = model.get_fit_model(dataset.features, dataset.targets);
+  silly_function_to_increment_stack_pointer();
+  const auto pred = fit_model.get_prediction(dataset.features);
+  silly_function_to_increment_stack_pointer();
 
-  auto cv_fast_scores =
-      cross_validated_scores(rmse, dataset, loo_indexers, model.get());
+  const Eigen::VectorXd pred_mean = pred.mean();
 
-  // Here we make sure the cross validated mean absolute error is reasonable.
-  // Note that because we are running leave one out cross validation, the
-  // RMSE for each fold is just the absolute value of the error.
-  EXPECT_LE(cv_scores.mean(), 0.1);
-}
+  const MarginalDistribution marginal = pred.marginal();
+  EXPECT_LE((pred_mean - marginal.mean).norm(), 1e-8);
```

Comment:
What is the type of `pred` here?
Is this testing that the predict types are consistent, or that conversion from a joint prediction to lower information forms is consistent.

Reply:
`pred` will be of type `Prediction<ModelType, Feature, FitType>` (see `prediction.h`).  The `Prediction` class is responsible for determining which `predict` methods are available for the given model, feature type and the way it was fit.  The "consistency" test is exactly what you mentioned.  That if you ask for a `JointDistribution` the mean and variance match what you'd get if you asked for the `MarginalDistribution` which has a mean identical to if you'd just asked for the mean.